File: codecamp/src/main/java/com/gdogaru/codecamp/view/session/SessionInfoFragment.java
/*
 * Copyright (C) 2008 Gabriel Dogaru (gdogaru@gmail.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.gdogaru.codecamp.view.session;


import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestOptions;
import com.gdogaru.codecamp.App;
import com.gdogaru.codecamp.R;
import com.gdogaru.codecamp.di.Injectable;
import com.gdogaru.codecamp.model.Session;
import com.gdogaru.codecamp.model.Speaker;
import com.gdogaru.codecamp.model.Track;
import com.gdogaru.codecamp.svc.CodecampClient;
import com.gdogaru.codecamp.util.DateUtil;
import com.gdogaru.codecamp.view.BaseFragment;
import com.google.firebase.analytics.FirebaseAnalytics;

import java.util.Locale;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import icepick.State;

public class SessionInfoFragment extends BaseFragment  implements Injectable {

    public static final String SESSION_ID = "sessionId";
    @BindView(R.id.sessionTitle)
    TextView sessionTitle;
    @BindView(R.id.sessionTime)
    TextView sessionTime;
    @BindView(R.id.sessionDescription)
    TextView sessionDescription;
    @BindView(R.id.sessionTrack)
    TextView sessionTrack;
    @BindView(R.id.speakerLayout)
    ViewGroup speakerLayout;
    @BindView(R.id.sessionTrackLayout)
    LinearLayout sessionTrackLayout;
    @BindView(R.id.speakerLayoutOuter)
    ViewGroup speakerLayoutOuter;
    @State
    String sessionId;

    @Inject
    CodecampClient codecampClient;
    @Inject
    FirebaseAnalytics firebaseAnalytics;

    public static SessionInfoFragment newInstance(String id) {
        SessionInfoFragment sessionInfoFragment = new SessionInfoFragment();
        sessionInfoFragment.setArguments(new Bundle());
        sessionInfoFragment.getArguments().putString(SESSION_ID, id);
        return sessionInfoFragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (savedInstanceState == null) {
            sessionId = getArguments().getString(SESSION_ID);
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.session_expanded_item, container, false);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        manage(ButterKnife.bind(this, view));

        initView();
    }

    public void initView() {
        Session session = codecampClient.getSession(sessionId);
        sessionTitle.setText(session.getTitle());
        sessionDescription.setText(session.getDescription());
        String timeString = DateUtil.formatPeriod(session.getStartTime(), session.getEndTime());
        sessionTime.setText(timeString);
        if (session.getTrack() != null) {
            Track track = codecampClient.getTrack(session.getTrack());
            if (track == null) {
                sessionTrack.setText("");
            } else {
                sessionTrack.setText(String.format(Locale.getDefault(), "%s, %s seats, %s", track.getName(), track.getCapacity(), track.getDescription()));
            }
        } else {
            sessionTrackLayout.setVisibility(View.GONE);
        }
        if (session.getSpeakerIds() == null || session.getSpeakerIds().size() == 0) {
            speakerLayoutOuter.setVisibility(View.GONE);
        } else {
            for (String id : session.getSpeakerIds()) {
                Speaker speaker = codecampClient.getSpeaker(id);
                if (speaker != null) {
                    addSpeaker(speaker);
                }
            }
        }

        Bundle bundle = new Bundle();
        bundle.putString(FirebaseAnalytics.Param.VALUE, session.getTitle());
        firebaseAnalytics.logEvent("session_view", bundle);
    }

    private void addSpeaker(Speaker speaker) {
        View speakerView = getActivity().getLayoutInflater().inflate(R.layout.session_speaker_info, speakerLayout, false);
        TextView speakerName = (TextView) speakerView.findViewById(R.id.speakerName);
        TextView speakerDesc = (TextView) speakerView.findViewById(R.id.speakerDescription);
        TextView company = (TextView) speakerView.findViewById(R.id.company);
        TextView job = (TextView) speakerView.findViewById(R.id.job_title);
        ImageView picture = (ImageView) speakerView.findViewById(R.id.speakerPhoto);

        speakerName.setText(speaker.getName());
        company.setText(speaker.getCompany());
        job.setText(speaker.getJobTitle());
        speakerDesc.setText(speaker.getBio());
        speakerLayout.addView(speakerView);

        Glide.with(getActivity())
                .load(speaker.getPhotoUrl())
                .apply(new RequestOptions()
                        .centerCrop()
                        .placeholder(R.drawable.person_icon))
//                .transition(withCrossFade(R.anim.fade_in))
                .into(picture);
    }
}

File: codecamp/src/main/java/com/gdogaru/codecamp/view/session/SessionInfoFragment.kt
/*
 * Copyright (c) 2019 Gabriel Dogaru - gdogaru@gmail.com
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *  *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package com.gdogaru.codecamp.view.session


import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import butterknife.BindView
import butterknife.ButterKnife
import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import com.gdogaru.codecamp.R
import com.gdogaru.codecamp.api.model.Session
import com.gdogaru.codecamp.api.model.Speaker
import com.gdogaru.codecamp.api.model.Track
import com.gdogaru.codecamp.di.Injectable
import com.gdogaru.codecamp.util.DateUtil
import com.gdogaru.codecamp.view.BaseFragment
import com.gdogaru.codecamp.view.MainViewModel
import com.google.firebase.analytics.FirebaseAnalytics
import java.util.*
import javax.inject.Inject

class SessionInfoFragment : BaseFragment(), Injectable {
    @BindView(R.id.sessionTitle)
    lateinit var sessionTitle: TextView
    @BindView(R.id.sessionTime)
    lateinit var sessionTime: TextView
    @BindView(R.id.sessionDescription)
    lateinit var sessionDescription: TextView
    @BindView(R.id.sessionTrack)
    lateinit var sessionTrack: TextView
    @BindView(R.id.speakerLayout)
    lateinit var speakerLayout: ViewGroup
    @BindView(R.id.sessionTrackLayout)
    lateinit var sessionTrackLayout: LinearLayout
    @BindView(R.id.speakerLayoutOuter)
    lateinit var speakerLayoutOuter: ViewGroup

    @Inject
    lateinit var firebaseAnalytics: FirebaseAnalytics
    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory
    private lateinit var viewModel: MainViewModel

    lateinit var sessionId: String

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewModel = ViewModelProviders.of(requireActivity(), viewModelFactory).get(MainViewModel::class.java)
        sessionId = arguments!!.getString(SESSION_ID)!!
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.session_expanded_item, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        manage(ButterKnife.bind(this, view))

        viewModel.getSession(sessionId).observe(this, androidx.lifecycle.Observer { s ->
            s?.let { initView(it.session, it.speakers, it.track) }
        })
    }

    fun initView(session: Session, speakers: List<Speaker>, track: Track?) {

        sessionTitle.text = session.title;
        sessionDescription.text = session.description;
        val timeString = DateUtil.formatPeriod(session.startTime, session.endTime)
        sessionTime.text = timeString

        if (track != null) {
            sessionTrack.text = String.format(Locale.getDefault(), "%s, %s seats, %s", track.name, track.capacity, track.description)
        } else {
            sessionTrack.text = ""
        }

        speakers.forEach { addSpeaker(it) }

        val bundle = Bundle().also {
            it.putString(FirebaseAnalytics.Param.VALUE, session.title)
        }
        firebaseAnalytics.logEvent("session_view", bundle);
    }

    private fun addSpeaker(speaker: Speaker) {
        val speakerView = requireActivity().layoutInflater.inflate(R.layout.session_speaker_info, speakerLayout, false)
        val speakerName = speakerView.findViewById<View>(R.id.speakerName) as TextView
        val speakerDesc = speakerView.findViewById<View>(R.id.speakerDescription) as TextView
        val company = speakerView.findViewById<View>(R.id.company) as TextView
        val job = speakerView.findViewById<View>(R.id.job_title) as TextView
        val picture = speakerView.findViewById<View>(R.id.speakerPhoto) as ImageView

        speakerName.text = speaker.name
        company.text = speaker.company
        job.text = speaker.jobTitle
        speakerDesc.text = speaker.bio
        speakerLayout.addView(speakerView)

        Glide.with(requireActivity())
                .load(speaker.photoUrl)
                .apply(RequestOptions()
                        .centerCrop()
                        .placeholder(R.drawable.person_icon))
                .into(picture)
    }

    companion object {

        const val SESSION_ID = "sessionId"

        fun newInstance(id: String): SessionInfoFragment {
            val sessionInfoFragment = SessionInfoFragment()
            sessionInfoFragment.arguments = Bundle()
            sessionInfoFragment.arguments!!.putString(SESSION_ID, id)
            return sessionInfoFragment
        }
    }
}
